[DE][Cypress] Re-enable cypress serverless tests subset

yctercero opened the pull request:
## Summary

Re-enable a subset of cypress serverless tests:

- Restructured `x-pack/test/security_solution_cypress/cypress/e2e/detection_response/rule_creation/custom_query_rule.cy.ts` to break it down into and enable the following tests:
    - `rule_creation/common_components.cy.ts`
    - `rule_creation/custom_query_rule.cy.ts`
    - `rule_details/common_components.cy.ts`
    - `rule_details/custom_query_rule.cy.ts`
    - `rule_edit/custom_query_rule.cy.ts`

yctercero committed:
re-enabling rule create common elements and custom query tests

yctercero committed:
figured out how to get the breadcrumb click to work

yctercero commented:
Flakey test runner - https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/3133

yctercero committed:
addres flake in rule details test

yctercero committed:
Merge branch 'main' of github.com:elastic/kibana into rule_create_gen

yctercero commented:
First flakey test runner showed some flake around one of the tests I re-enabled. Pushed up a fix and trying again:

https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/3138

yctercero committed:
skipping deletion test for now

yctercero committed:
update codeowners

yctercero commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into rule_create_gen

banderror commented on x-pack/test/security_solution_cypress/cypress/e2e/detection_response/rule_creation/common_components.cy.ts:
```diff
@@ -0,0 +1,101 @@
+import { login, visit } from '../../../tasks/login';
+import { RULE_CREATION } from '../../../urls/navigation';
+
+// This test is meant to test touching all the common various components in rule creation
+// to ensure we don't miss any changes that maybe affect one of these more obscure UI components
+// in the creation form. For any rule type specific functionalities, please include
+// them in the relevant /rule_creation/[RULE_TYPE].cy.ts test.
+describe('Common rule creation components', { tags: ['@ess', '@serverless'] }, () => {
```
Since this is not skipped, do we know for sure this test is not flaky in Serverless?

banderror commented on x-pack/test/security_solution_cypress/cypress/e2e/detection_response/rule_details/custom_query_rule.cy.ts:
```diff
@@ -0,0 +1,55 @@
+    deleteAlertsAndRules();
+    createRule(rule);
+    login();
+    visitWithoutDateRange(SECURITY_DETECTIONS_RULES_URL);
+    getRulesManagementTableRows().should('have.length', 1);
+  });
+
+  it('Displays rule details', function () {
+    goToRuleDetailsOf(rule.name);
```
Let's not test navigation from the Rule Management to the Rule Details page in this test, I think navigation should be tested separately. We can open the Rule Details page right away.

banderror commented on x-pack/test/security_solution_cypress/kubectl:
👀

banderror reviewed (changes requested):
@yctercero Let's not commit the `x-pack/test/security_solution_cypress/kubectl` binary 🙂 

The rest looks good, I mainly reviewed changes for the Rule Management team and left some suggestions. Thank you for creating separate tests for rule deletion and the details page!

My main concern is how we ensure that the tests we're enabling in this PR are not flaky in Serverless. What's our plan?

yctercero commented on x-pack/test/security_solution_cypress/cypress/e2e/detection_response/rule_management/rule_actions/deletion/rule_delete.cy.ts:
```diff
@@ -0,0 +1,102 @@
+    rule_id: 'rule1',
+    name: 'New Rule Test',
+    enabled: false,
+    max_signals: 500,
```
I don't believe so. I just moved tests over, didn't spend time on this file in particular since I wasn't yet trying to enable this suite. But I'll update with your suggestions!

yctercero commented on x-pack/test/security_solution_cypress/cypress/e2e/detection_response/rule_management/rule_actions/deletion/rule_delete.cy.ts:
```diff
@@ -0,0 +1,102 @@
+      selectRulesByName(rulesToDelete);
+      deleteSelectedRules();
+
+      getRulesManagementTableRows()
+        .first()
+        .within(() => {
+          cy.get(RULE_SWITCH).should('not.exist');
+        });
+
+      getRulesManagementTableRows().should('have.length', expectedNumberOfRulesAfterDeletion);
+      findAllRules().then(({ body }) => {
+        const numberOfRules = body.data.length;
+        expect(numberOfRules).to.eql(expectedNumberOfRulesAfterDeletion);
+      });
+      getRulesManagementTableRows()
+        .first()
+        .within(() => {
+          cy.get(RULE_SWITCH).should('exist');
+        });
```
Not sure 😅 Moved these over from `custom_query_rule.cy.ts`. I'll take a look and add comments here.

yctercero committed:
updating tests per feedback

yctercero committed:
Merge branch 'rule_create_gen' of github.com:yctercero/kibana into rule_create_gen

yctercero committed:
Merge branch 'main' of github.com:elastic/kibana into rule_create_gen

yctercero commented:
Thanks for the review @banderror ! I addressed your concerns. As for ensuring they're not flakey on serverless. We don't yet have them running in the flakey test runner so that is not an option. I did try running them locally many times over and increased the burn rate to 25. The burn scripts are running for serverless in CI, so that should also provide us some confidence.

banderror commented on x-pack/test/security_solution_cypress/cypress/tasks/create_new_rule.ts:
```diff
@@ -126,11 +127,15 @@ import { waitForAlerts } from './alerts';
 import { refreshPage } from './security_header';
 import { EMPTY_ALERT_TABLE } from '../screens/alerts';
 
+export const createAndEnableRuleOnly = () => {
+  cy.get(CREATE_AND_ENABLE_BTN).click({ force: true });
+  cy.get(CREATE_AND_ENABLE_BTN).should('not.exist');
+};
```
Posted a comment on the `createAndEnableRule` function in https://github.com/elastic/kibana/pull/166767#discussion_r1330689150 - I guess you had to create this new function because the existing one was doing navigation using breadcrumbs.

banderror reviewed (approved):
Alright, let's do it and see what happens 🤞

maximpn commented on x-pack/test/security_solution_cypress/cypress/tasks/create_new_rule.ts:
```diff
@@ -126,11 +127,15 @@ import { waitForAlerts } from './alerts';
 import { EMPTY_ALERT_TABLE } from '../screens/alerts';
 
+export const createAndEnableRuleOnly = () => {
+  cy.get(CREATE_AND_ENABLE_BTN).click({ force: true });
```
It should be safe to omit `{ force: true }`.

yctercero replied:
Good call. I'll update.

yctercero commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into rule_create_gen

yctercero commented:
Running flakey test runner for serverless. Thanks @maximpn ! 

https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/3193

vitaliidm commented on x-pack/test/security_solution_cypress/cypress/e2e/detection_response/rule_details/common_components.cy.ts:
```diff
@@ -0,0 +1,179 @@
+// This test is meant to test all common aspects of the rule details page that should function
+// the same regardless of rule type. For any rule type specific functionalities, please include
+// them in the relevant /rule_details/[RULE_TYPE].cy.ts test.
+describe('Common components', { tags: ['@ess', '@serverless'] }, () => {
```
nit: maybe renaming to common scenarios would slightly better fit to its purpose?
Since we test rule types scenarios, not isolated components

vitaliidm commented on x-pack/test/security_solution_cypress/cypress/e2e/detection_response/rule_management/rule_actions/deletion/rule_delete.cy.ts:
```diff
@@ -0,0 +1,88 @@
+import { deleteAlertsAndRules } from '../../../../../tasks/common';
+import { login, visitSecurityDetectionRulesPage } from '../../../../../tasks/login';
+
+describe('Rule deletion', { tags: ['@ess', '@serverless', '@skipInServerless'] }, () => {
```
rule deletion still should be skipped in serverless?

vitaliidm commented on x-pack/test/security_solution_cypress/cypress/tasks/create_new_rule.ts:
```diff
@@ -294,6 +299,23 @@ export const fillReferenceUrls = (referenceUrls: string[] = ruleFields.reference
+};
+
+export const fillAboutRuleMinimumAndContinue = (rule: RuleCreateProps) => {
+  cy.get(RULE_NAME_INPUT).clear({ force: true });
```
similar to a comment above, do we need `force:true` everywhere?

vitaliidm commented:
>  I did try running them locally many times over and increased the burn rate to 25. The burn scripts are running for serverless in CI, so that should also provide us some confidence.

@Yara, I wasn't able to find serverless burn specs in build, only ESS https://buildkite.com/elastic/kibana-pull-request/builds/160419#_

vitaliidm commented on x-pack/test/security_solution_cypress/cypress/e2e/detection_response/rule_creation/common_components.cy.ts:
```diff
@@ -0,0 +1,100 @@
+
+    createAndEnableRuleOnly();
+
+    cy.get(RULE_NAME_HEADER).should('contain', ruleFields.ruleName);
```
now, that we don't check if rule details displayed, after rule was created, should we introduce one test case on common scenarios for this?
So, we can be sure, that after user creates rule, they get redirected to details page and nothing is broken in this workflow

yctercero replied:
If it's ok, I was able to clean up and add some scenarios to test details, but it's added a lot more changes. If it's ok, I'd prefer to put this PR in as is as it's confirmed to be passing consistently and then follow up with more changes/test cases.

maximpn commented on x-pack/test/security_solution_cypress/cypress/e2e/detection_response/rule_creation/common_components.cy.ts:
```diff
@@ -0,0 +1,100 @@
+    login();
+  });
+
+  it('Creates and enables a rule', function () {
```
nit: Generally speaking we don't have to enable to rule to create it. Ideally we should test both `create enabled` and `create disabled`. I think it's ok to leave it as is and do the refactor later on in follow up PRs.

maximpn reviewed (approved):
@yctercero thank you for improving custom query rule Cypress tests 🙏

I have a few general thoughts about rule creation tests (all rule types). As a lot of things can fail and the test should check the specified in create rule form parameters are reflected on the rule details page (ideally in rule API response as well). Achieving that state of art coverage will take time should be addressed separately. It'd be logical to have a test plan first and then extend the tests according to it.

This PR is a good step towards concerns separation and better coverage so approving it. Any kind of improvements we should address later on.

yctercero committed:
Merge branch 'main' of github.com:elastic/kibana into rule_create_gen

vitaliidm reviewed (approved):
Thanks for working on this PR

I like the idea of splitting it into common flows, so we can assert every property in rule flow.
And then, having a separate for rule type - so rule specific fields/scenarios can be tested

yctercero committed:
updating refs after merging with main

yctercero committed:
Merge branch 'main' of github.com:elastic/kibana into rule_create_gen

yctercero committed:
fix references

yctercero committed:
update edit rule test

yctercero commented:
Running cypress serverless flakey test runner once more.

https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/3219

yctercero committed:
Delete x-pack/test/security_solution_cypress/cypress/e2e/detection_response/rule_creation/blob.cy.ts

yctercero commented:
Again after removing extra test file - https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/3220

yctercero commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into rule_create_gen

yctercero commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into rule_create_gen

yctercero commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into rule_create_gen

yctercero commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into rule_create_gen

yctercero commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into rule_create_gen

yctercero commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into rule_create_gen

yctercero commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into rule_create_gen
